adds concurrency options to the async package

kaperys opened the pull request:
~~Adds a `simple` wrapper which provides a smaller and simpler api for substrate sources.~~

Adds concurrency options to the `async` package.

In the scenario a user simply wants to consumer messages from a source, and for whatever reason does not care about acks (for example if they're using a source that doesn't care about them or they're using something such as the `ackordering` pkg) the current substrate api forces the user to implement boilerplate code (channel and context management, concurrency, etc) to get something up and running. The `simple` package removes the requirement for such code, offering a much smaller, but simpler api.

kaperys committed:
added simple wrapper

kaperys committed:
go mod tidy

heedson commented:
@kaperys What makes this different from the `async` package in this repo? If there's additional functionality, maybe we can work towards extending the `async` one? I think `simple` is a bit vague for a package name in this case. Please give me your thoughts!
https://github.com/uw-labs/substrate-tools/tree/master/async

As a note, `async` emulates the interface style that the sync wrapper uses in substrate itself. i.e. having the handler passed at the `Consume` (or `Run`) method:
https://github.com/uw-labs/substrate/blob/master/sync_adapter_source.go

kaperys commented:
@heedson I hadn't actually realised the `async` package provided the functionality it does - my apologies. It looks like it does mostly the same thing. What is the intended method of controlling the concurrency of the executing the `handler` function - https://github.com/uw-labs/substrate-tools/blob/master/async/source.go#L72. Is the user expected to concurrently call `MessageSource.ConsumeMessages`?

heedson commented:
@kaperys and I had a discussion "offline" and decided that we can move the variadic options to the `async` package to extend current functionality.

kaperys committed:
added concurrency support to async

kaperys committed:
test: don't cancel context on ack

heedson commented on async/source.go:
```diff
@@ -32,59 +56,71 @@ type MessageSource interface {
+	source    substrate.AsyncMessageSource
+	msgBuffer int
+	ackBuffer int
+	consumers int
```
Perhaps these should be `uint` or the `WithXXX` should have some sort of validation baked in? -1 consumers or `ackBuffer` doesn't sound healthy! :smile:

heedson reviewed (changes requested):
LGTM - just a single validation comment about your additions. It looks like you may have also added some README updates that are unrelated?

kaperys committed:
buffer sizes as uint, adds examples, update readme

kaperys commented:
@heedson The README updates are just an indentation change on the flushing example code - not sure why GitHub is showing the changes as it is. I've changed the buffer and consumer sizes to `uint` and added some example code. Would you mind taking another look, please?

heedson reviewed (dismissed):
LGTM

povilasv commented on async/source.go:
```diff
@@ -32,59 +56,71 @@ type MessageSource interface {
 
-	messages := make(chan substrate.Message)
-	acks := make(chan substrate.Message)
+	messages := make(chan substrate.Message, a.msgBuffer)
```
I think by default this would be 0, so it will fail, maybe we can change defaults to 1?

kaperys replied:
The default is 0, yes, which means the channel will be unbuffered https://golang.org/pkg/builtin/#make

heedson replied:
0 will make a blocking, unbuffered channel. The tests show it works with the default of 0.

povilasv replied:
ah ok, makes sense, ty for explanation

povilasv commented on async/source_test.go:
```diff
@@ -66,11 +68,65 @@ func TestConsumeMessagesSuccessfully(t *testing.T) {
+	go func() {
+		defer close(errs)
+		errs <- source.ConsumeMessages(sourceContext, func(_ context.Context, _ substrate.Message, ack AckFunc) error {
+			wg.Done()
```
shouldn't we be adding to waitgroup for this goroutine?

povilasv replied:
I mean we are adding waitgroup in https://github.com/uw-labs/substrate-tools/pull/6/files#diff-90006bcabbe2b6ae8a5c9a3d4157a6dbR80 but calling wg.Done() in different one, which is confusing me a bit.

povilasv replied:
example usage https://gobyexample.com/waitgroups does this for the same goroutine

povilasv replied:
even better in the example usage go docs https://golang.org/pkg/sync/#WaitGroup

kaperys replied:
I've moved the `wg.Add()` call into the main routine for better readability.

povilasv commented on examples/async/async/main.go:
```diff
@@ -0,0 +1,55 @@
+	defer cancel()
+
+	handler := func(ctx context.Context, msg substrate.Message, ack async.AckFunc) error {
+		defer ack()
+
+		println("consumed:", string(msg.Data()))
```
```suggestion
		println("consumed:", string(msg.Data()))
              ack()
```

I think let's simplify to this, wdyt?

povilasv replied:
also handle err from ack function, so maybe:

if err := ack(); err !=nil {
return err.
}

kaperys replied:
As this is example code, I image most people will copy and paste it. Therefore I prefer to defer the `ack()` right at the top of the handler code - that way there's less chance of it being removed when modified resulting in broken code.

heedson replied:
I agree with @kaperys. I appreciate an example being the bare bones to show it in a simple solution, but having it deferred is a more extendable example that people can look towards. As soon as there is branching in the handler func (if someone uses the example as a starting base), then the `ack()` will be moved. When implementing the ack func, the intention was to have it used like this in the default use cases.

povilasv replied:
Ok I agree with defer, but then maybe let's add some error handling

kaperys replied:
@povilasv and I had a chat offline and agreed that `return ack()` at the end of the function is probably a good approach. This way if the users message handler code returns an error and the ack() function returns an error (or, would have returned an error), the users message handler code error is returned first. What do you think @heedson?

heedson replied:
That's fine, but for the copy and paste use case, this will result in the same situation for the user. As soon as a 2nd `return` is added, there will be a slight bit of management added. Either is fine for an example so I'm comfortable with this.

I think it maybe is worth us/myself looking at the ack function signature and seeing if it makes sense without an error returning so this isn't even an issue in the future.

kaperys replied:
Great! I agree that if the ack function didn't return an error orchestrating this would be simpler.

povilasv commented on examples/async/sync/main.go:
```diff
@@ -0,0 +1,46 @@
+	defer cancel()
+
+	handler := func(ctx context.Context, msg substrate.Message, ack async.AckFunc) error {
+		defer ack()
```
ditto

kaperys replied:
Will simplify.

povilasv commented on examples/async/async/main.go:
```diff
@@ -0,0 +1,55 @@
+		}
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
do we need a timeoutable context in this example? maybe we can simplify it?

kaperys replied:
Will simplify.

povilasv commented on examples/async/sync/main.go:
```diff
@@ -0,0 +1,46 @@
+		}
+	}()
+
+	ctx, cancel := context.WithTimeout(context.Background(), 5*time.Second)
```
ditto

kaperys replied:
Will simplify.

povilasv commented on async/source_test.go:
```diff
@@ -66,11 +68,65 @@ func TestConsumeMessagesSuccessfully(t *testing.T) {
+		ackBuffer: 100,
+	}
+
+	sourceContext, sourceCancel := context.WithTimeout(context.Background(), time.Second)
```
do we really need a timeoutable context in this test case?

heedson replied:
I find time out contexts allow the tests to show that the test code works within a reasonable window and don't hang.

kaperys replied:
Simply following the pattern in the rest of the tests. I would verge on the side of leaving this, however. It means broken tests won't hang indefinitely.

povilasv replied:
ok, ty for clarification

povilasv commented on async/source.go:
```diff
@@ -32,59 +56,71 @@ type MessageSource interface {
 
 func (a *messageSourceAdapter) Close() error {
-	return a.ac.Close()
+	return a.source.Close()
```
I feel like we ought to shutdown goroutines launched in `ConsumeMessages` first, wdyt?

povilasv replied:
discussed privately, but this will just work

kaperys committed:
simplify example code, fix concurrent test

kaperys committed:
merge upstream/master

kaperys commented:
@heedson Are you happy with these changes? Should I merge?

heedson reviewed (approved):
LGTM
